### PR TITLE
feat(npm): use Yarn 3 mode to skip install or build

### LINKED ISSUE #7584: Don't change pnp package location from unplugged to cache
**What Renovate type, platform and version are you using?**

Self hosted

**Describe the bug**

Using yarn 2, renovate changes the package location from unplugged to cache in the .pnp.js file.

Running yarn locally then changes it back again. Which means this file needs to be recommitted after every PR.

image

**To Reproduce**

This only happens with packages that need to be unplugged by yarn 2. `nodemon` for example.

### LINKED ISSUE #10061: Yarn 2 install timeout causes Temporary error / abort
**Renovate type**

- [x] Hosted app
- [ ] Self hosted

**Renovate platform**

- [ ] Azure DevOps (dev.azure.com)
- [ ] Azure DevOps Server (self-hosted)
- [ ] Bitbucket Cloud (bitbucket.org)
- [ ] Bitbucket Server (self-hosted)
- [ ] Gitea (gitea.com, self-hosted)
- [x] GitHub (github.com, GitHub Enterprise)
- [ ] GitLab (gitlab.com, self-hosted)

**Describe the bug**

Renovate won't run - logs end with `Temporary error - aborting`

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
DEBUG: Removing container(branch="renovate/sentry-monorepo")
{
  "containerId": "d8d5f9bde905"
}
DEBUG: exec interrupted by SIGTERM - run needs to be aborted(branch="renovate/sentry-monorepo")
{
  "err": {
    "killed": true,
    "code": null,
    "signal": "SIGTERM",
    "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/gh/env0/env0\":\"/tmp/renovate/gh/env0/env0\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_ENABLE_IMMUTABLE_INSTALLS -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/tmp/renovate/gh/env0/env0\" docker.io/renovate/node bash -l -c \"npm i -g yarn && yarn install\"",
    "stdout": "\n> yarn@1.22.10 preinstall /home/ubuntu/.npm-global/lib/node_modules/yarn\n> :; (node ./preinstall.js > /dev/null 2>&1 || true)\n\n/home/ubuntu/.npm-global/bin/yarn -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n/home/ubuntu/.npm-global/bin/yarnpkg -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n+ yarn@1.22.10\nadded 1 package in 1.692s\n➤ YN0000: ┌ Project validation\n➤ YN0057: │ @env0/parent: 'nohoist' is deprecated, please use 'installConfig.hoistingLimits' instead\n➤ YN0000: └ Completed\n➤ YN0000: ┌ Resolution step\n➤ YN0002: │ @ant-design/icons@npm:4.6.2 [ac7b5] doesn't provide react-dom (pa8f5f), requested by rc-util\n➤ YN0002: │ @env0/common-appsync-push@workspace:common/appsync-push doesn't provide @react-native-community/async-storage (p695a4), requested by aws-appsync\n➤ YN0002: │ @env0/common-appsync-push@workspace:common/appsync-push doesn't provide @react-native-community/netinfo (p5e3e9), requested by aws-appsync\n➤ YN0002: │ @env0/common-appsync-schema@workspace:common/appsync-schema doesn't provide graphql (p9d29c), requested by graphql-tag\n➤ YN0002: │ @env0/common-dev@workspace:common/dev doesn't provide jest (pe40f6), requested by ts-jest\n➤ YN0060: │ @env0/common-dev@workspace:common/dev provides serverless (p5b208) with version 2.41.2, which doesn't satisfy what serverless-plugin-monorepo requests\n➤ YN0002: │ @env0/common-dev@workspace:common/dev doesn't provide typescript (pe1342), requested by fork-ts-checker-webpack-plugin\n➤ YN0002: │ @env0/common-dev@workspace:common/dev doesn't provide typescript (p709c8), requested by ts-jest\n➤ YN0002: │ @env0/common-dev@workspace:common/dev doesn't provide typescript (p5c0f3), requested by ts-loader\n➤ YN0060: │ @env0/common-dev@workspace:common/dev provides webpack (pd03b9) with version 5.37.0, which doesn't satisfy what cache-loader requests\n➤ YN0060: │ @env0/common-dev@workspace:common/dev provides webpack (p70294) with version 5.37.0, which doesn't satisfy what thread-loader requests\n➤ YN0002: │ @env0/common-dynamo@workspace:common/dynamo doesn't provide aws-sdk (pa308a), requested by @shelf/jest-dynamodb\n➤ YN0002: │ @env0/common-dynamo@workspace:common/dynamo doesn't provide aws-sdk (pf3e2f), requested by @shiftcoders/dynamo-easy\n➤ YN0002: │ @env0/common-dynamo@workspace:common/dynamo doesn't provide reflect-metadata (p180a1), requested by @shiftcoders/dynamo-easy\n➤ YN0002: │ @env0/common-dynamo@workspace:common/dynamo doesn't provide tslib (p8218b), requested by @shiftcoders/dynamo-easy\n➤ YN0002: │ @env0/common-lambda@workspace:common/lambda doesn't provide typescript (p23a0c), requested by tsargs\n➤ YN0002: │ @env0/common-rds@workspace:common/rds doesn't provide @types/bluebird (p0d0cc), requested by sequelize-typescript\n➤ YN0002: │ @env0/common-rds@workspace:common/rds doesn't provide @types/node (p23ed1), requested by sequelize-typescript\n➤ YN0002: │ @env0/common-rds@workspace:common/rds doesn't provide @types/validator (p4a337), requested by sequelize-typescript\n➤ YN0002: │ @env0/common-rds@workspace:common/rds doesn't provide reflect-metadata (p6e6f6), requested by sequelize-typescript\n➤ YN0060: │ @env0/deployment-service@workspace:services/deployment-service provides webpack (p2d9e4) with version 5.37.0, which doesn't satisfy what webpack-cli requests\n➤ YN0002: │ @env0/frontend@workspace:frontend doesn't provide @nivo/tooltip (p6d783), requested by @nivo/core\n➤ YN0002: │ @env0/frontend@workspace:frontend doesn't provide @react-native-community/async-storage (pa7987), requested by aws-appsync\n➤ YN0002: │ @env0/frontend@workspace:frontend doesn't provide @react-native-community/netinfo (p051ae), requested by aws-appsync\n➤ YN0002: │ @env0/frontend@workspace:frontend doesn't provide prop-types (pdbb2b), requested by @nivo/core\n➤ YN0002: │ @env0/frontend@workspace:frontend doesn't provide prop-types (pa3898), requested by @nivo/line\n➤ YN0002: │ @env0/frontend@workspace:frontend doesn't provide prop-types (p5ee6d), requested by react-scroll-into-view-if-needed\n➤ YN0060: │ @env0/frontend@workspace:frontend provides typescript (p3c8fd) with version 3.9.9, which doesn't satisfy what react-intl and some of its descendants request\n➤ YN0002: │ @env0/frontend@workspace:frontend doesn't provide webpack (pa45bf), requested by speed-measure-webpack-plugin\n➤ YN0002: │ @env0/organization-service@workspace:services/organization-service doesn't provide serverless (pec91f), requested by serverless-plugin-git-variables\n➤ YN0002: │ @env0/parent@workspace:. doesn't provide ts-node (p72795), requested by ttypescript\n➤ YN0002: │ @env0/self-hosted-agent@workspace:self-hosted-agent doesn't provide serverless (p91f3e), requested by serverless-plugin-monorepo\n➤ YN0002: │ @env0/self-hosted-agent@workspace:self-hosted-agent doesn't provide ts-node (pd099a), requested by ttypescript\n➤ YN0002: │ @env0/self-hosted-agent@workspace:self-hosted-agent doesn't provide typescript (pdb2dd), requested by ttypescript\n➤ YN0002: │ @env0/self-hosted-agent@workspace:self-hosted-agent doesn't provide webpack (p5d645), requested by serverless-webpack\n➤ YN0002: │ @env0/shared@workspace:shared doesn't provide webpack (p14d5f), requested by copy-webpack-plugin\n➤ YN0002: │ @serverless/components@npm:3.9.2 doesn't provide inquirer (p497c3), requested by inquirer-autocomplete-prompt\n➤ YN0002: │ antd@npm:4.15.6 [ac7b5] doesn't provide dayjs (p14bc9), requested by rc-picker\n➤ YN0002: │ apollo-link-context@npm:1.0.11 doesn't provide graphql (p3926d), requested by apollo-link\n➤ YN0002: │ apollo-link-dedup@npm:1.0.20 doesn't provide graphql (p1aa56), requested by apollo-link\n➤ YN0002: │ apollo-link-error@npm:1.1.13 doesn't provide graphql (p7d98f), requested by apollo-link\n➤ YN0002: │ apollo-link-retry@npm:2.2.7 doesn't provide graphql (pec3bc), requested by apollo-link\n➤ YN0002: │ aws-appsync-auth-link@npm:2.0.3 doesn't provide graphql (p0c003), requested by apollo-link\n➤ YN0002: │ aws-appsync-subscription-link@npm:2.2.1 [65954] doesn't provide graphql (p42908), requested by apollo-link\n➤ YN0002: │ aws-appsync-subscription-link@npm:2.2.1 [65954] doesn't provide graphql (p33073), requested by apollo-link-http\n➤ YN0002: │ craco-babel-loader@npm:0.1.4 doesn't provide react-scripts (p7fc30), requested by @craco/craco\n➤ YN0002: │ craco-less@npm:1.17.0 [4dab7] doesn't provide webpack (p014e7), requested by less-loader\n➤ YN0002: │ rc-mentions@npm:1.5.2 [b0704] doesn't provide react-dom (p47b2e), requested by rc-menu\n➤ YN0002: │ rc-mentions@npm:1.5.2 [b0704] doesn't provide react-dom (p7e71b), requested by rc-textarea\n➤ YN0002: │ rc-menu@npm:8.7.1 doesn't provide react (pa3843), requested by rc-motion\n➤ YN0002: │ rc-menu@npm:8.7.1 doesn't provide react (pbfc9e), requested by mini-store\n➤ YN0002: │ rc-menu@npm:8.7.1 doesn't provide react-dom (p994eb), requested by rc-motion\n➤ YN0002: │ rc-menu@npm:8.7.1 doesn't provide react-dom (p68b1b), requested by mini-store\n➤ YN0002: │ rc-trigger@npm:5.0.6 doesn't provide react (pbe38e), requested by rc-align\n➤ YN0002: │ rc-trigger@npm:5.0.6 doesn't provide react (p523b8), requested by rc-motion\n➤ YN0002: │ rc-trigger@npm:5.0.6 doesn't provide react-dom (p7d956), requested by rc-align\n➤ YN0002: │ rc-trigger@npm:5.0.6 doesn't provide react-dom (p641b1), requested by rc-motion\n➤ YN0002: │ rc-trigger@npm:5.2.0 doesn't provide react (pf9157), requested by rc-align\n➤ YN0002: │ rc-trigger@npm:5.2.0 doesn't provide react (pd23c7), requested by rc-motion\n➤ YN0002: │ rc-trigger@npm:5.2.0 doesn't provide react (pa2834), requested by rc-util\n➤ YN0002: │ rc-trigger@npm:5.2.0 doesn't provide react-dom (p963e4), requested by rc-align\n➤ YN0002: │ rc-trigger@npm:5.2.0 doesn't provide react-dom (p1bd2c), requested by rc-motion\n➤ YN0002: │ rc-trigger@npm:5.2.0 doesn't provide react-dom (p67b6b), requested by rc-util\n➤ YN0002: │ react-dev-utils@npm:10.2.1 doesn't provide typescript (p62ad2), requested by fork-ts-checker-webpack-plugin\n➤ YN0002: │ react-dev-utils@npm:10.2.1 doesn't provide webpack (p2e56c), requested by fork-ts-checker-webpack-plugin\n➤ YN0002: │ react-show-more@npm:2.0.0 [ac7b5] doesn't provide prop-types (p8c343), requested by react-truncate\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide @react-three/fiber (pec093), requested by @react-spring/three\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide konva (p5fe80), requested by @react-spring/konva\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react (p3099c), requested by @react-spring/core\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react (pc44d5), requested by @react-spring/konva\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react (p288bd), requested by @react-spring/native\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react (pb5fdd), requested by @react-spring/three\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react (p603d2), requested by @react-spring/web\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react (pea741), requested by @react-spring/zdog\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react-dom (pb8433), requested by @react-spring/web\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react-dom (p2b8c0), requested by @react-spring/zdog\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react-konva (p82726), requested by @react-spring/konva\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react-native (p99088), requested by @react-spring/native\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide react-zdog (pf4dce), requested by @react-spring/zdog\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide three (p22c55), requested by @react-spring/three\n➤ YN0002: │ react-spring@npm:9.1.2 doesn't provide zdog (p51736), requested by @react-spring/zdog\n➤ YN0002: │ serverless-webpack@npm:5.5.0 [03f21] doesn't provide typescript (p79574), requested by ts-node\n➤ YN0002: │ serverless-webpack@npm:5.5.0 [f955a] doesn't provide typescript (p33c9f), requested by ts-node\n➤ YN0002: │ typewriter@npm:7.4.1 doesn't provide @types/react (p0cd90), requested by ink\n➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code\n➤ YN0000: └ Completed in 3s 443ms\n➤ YN0000: ┌ Fetch step\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT No package manager detected; defaulting to Yarn\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT \n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT No package manager detected; defaulting to Yarn\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT \n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT ➤ YN0000: LICENSE\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT ➤ YN0000: README.md\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT ➤ YN0000: example/index.js\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT ➤ YN0000: lib/index.js\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT ➤ YN0000: package.json\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT ➤ YN0000: test/index.js\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT ➤ YN0000: Package archive generated in /tmp/xfs-0a892fa0/package.tgz\n➤ YN0000: │ /tmp/xfs-0a892fa0 STDOUT ➤ YN0000: Done in 0s 414ms\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT ➤ YN0000: LICENSE\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT ➤ YN0000: README.md\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT ➤ YN0000: example/index.js\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT ➤ YN0000: lib/index.js\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT ➤ YN0000: package.json\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT ➤ YN0000: test/index.js\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT ➤ YN0000: Package archive generated in /tmp/xfs-a4e6fcf9/package.tgz\n➤ YN0000: │ /tmp/xfs-a4e6fcf9 STDOUT ➤ YN0000: Done in 0s 368ms\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT No package manager detected; defaulting to Yarn\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT \n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: CONTRIBUTING.md\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: DOCUMENTATION.md\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: LICENSE\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: README.md\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: example/index.js\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: lib/index.js\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: package.json\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: test/index.js\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: Package archive generated in /tmp/xfs-31938b6e/package.tgz\n➤ YN0000: │ /tmp/xfs-31938b6e STDOUT ➤ YN0000: Done in 0s 228ms\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT No package manager detected; defaulting to Yarn\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT \n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .dockerignore\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .gitlab-ci.yml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .idea/cfn-mysql-user-provider.iml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .idea/inspectionProfiles/Project_Default.xml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .idea/misc.xml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .idea/modules.xml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .idea/vcs.xml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .make-release-support\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: .release\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: Dockerfile.lambda\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: LICENSE\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: Makefile\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: Makefile.mk\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: README.md\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: cloudformation/cfn-resource-provider.yaml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: cloudformation/complete-stack.yaml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: cloudformation/demo-stack.yaml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: cloudformation/vpc.yaml\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: docs/MySQLUser.md\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: package.json\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: requirements.txt\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: src/mysql_user_provider.py\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: test-requirements.txt\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: tests/README.md\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: tests/test_mysql_user_provider.py\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: Package archive generated in /tmp/xfs-13f53abc/package.tgz\n➤ YN0000: │ /tmp/xfs-13f53abc STDOUT ➤ YN0000: Done in 1s 18ms\n➤ YN0000: │ /tmp/xfs-ce300955 STDOUT No package manager detected; defaulting to Yarn\n➤ YN0000: │ /tmp/xfs-ce300955 STDOUT \n➤ YN0000: │ /tmp/xfs-e25a07dd STDOUT Installing the project using Yarn Classic\n➤ YN0000: │ /tmp/xfs-e25a07dd STDOUT \n➤ YN0000: │ /tmp/xfs-ce300955 STDOUT ➤ YN0000: README.md\n➤ YN0000: │ /tmp/xfs-ce300955 STDOUT ➤ YN0000: package.json\n➤ YN0000: │ /tmp/xfs-ce300955 STDOUT ➤ YN0000: src/serverless-environment-to-key-value-pair-plugin.js\n➤ YN0000: │ /tmp/xfs-ce300955 STDOUT ➤ YN0000: Package archive generated in /tmp/xfs-ce300955/package.tgz\n➤ YN0000: │ /tmp/xfs-ce300955 STDOUT ➤ YN0000: Done in 0s 359ms\n➤ YN0000: │ /tmp/xfs-e25a07dd STDOUT ➤ YN0000: Downloading https://**redacted**@patch:fsevents@npm%3A2.1.2#builtin<compat/fsevents>::version=2.1.2&hash=11e9ea The platform linux is incompatible with this module, link skipped.\n➤ YN0062: │ fsevents@patch:fsevents@npm%3A2.3.2#builtin<compat/fsevents>::version=2.3.2&hash=11e9ea The platform linux is incompatible with this module, link skipped.\n➤ YN0062: │ fsevents@patch:fsevents@npm%3A1.2.9#builtin<compat/fsevents>::version=1.2.9&hash=11e9ea The platform linux is incompatible with this module, link skipped.\n",
    "stderr": "",
    "message": "Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/gh/env0/env0\":\"/tmp/renovate/gh/env0/env0\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_ENABLE_IMMUTABLE_INSTALLS -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/tmp/renovate/gh/env0/env0\" docker.io/renovate/node bash -l -c \"npm i -g yarn && yarn install\"\n",
    "stack": "Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/gh/env0/env0\":\"/tmp/renovate/gh/env0/env0\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_ENABLE_IMMUTABLE_INSTALLS -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/tmp/renovate/gh/env0/env0\" docker.io/renovate/node bash -l -c \"npm i -g yarn && yarn install\"\n\n    at ChildProcess.exithandler (child_process.js:308:12)\n    at ChildProcess.emit (events.js:315:20)\n    at ChildProcess.EventEmitter.emit (domain.js:467:12)\n    at maybeClose (internal/child_process.js:1048:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
  }
}
DEBUG: Passing TEMPORARY_ERROR error up(branch="renovate/sentry-monorepo")
INFO: Temporary error - aborting
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

We've upgraded to yarn v2 yesterday - so that may have something to do with that?...  
Please let us know if there are any more details we can share to help pinpoint the root cause

### PR DESCRIPTION
**Pull request by ylemkimon**
## Changes:

* Set `skipInstalls = false` if [zero-installs](https://yarnpkg.com/features/zero-installs) is used.
* Use new Yarn 3 `--mode=update-lockfile` flag to not run the link step and only fetch what's necessary to compute an updated lockfile if `config.skipInstalls !== false`.
* Use `--mode=skip-build` to not run the build scripts instead of skipping the whole build step (`enableScripts: false`) if `!adminConfig.allowScripts || config.ignoreScripts`.

## Context:

This PR fixes #10061 and reduces network bandwidth by fetching only necessary packages. This PR also fixes #7584 by running the build step that unplugs packages.

### Caveat

Yarn built between https://github.com/yarnpkg/berry/commit/9bcd27ae34aee77a567dd104947407532fa179b3 and https://github.com/yarnpkg/berry/commit/d13213280191f064785cf0bc19c92c76c60581bd would cause an invalid option error. This would affect:
  - Yarn 3 release candidates (`3.0.0-rc.1` to `3.0.0-rc.9`)
  - `yarn set version from sources` run between two commits

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository (https://github.com/ylemkimon/renovate-yarn-3)

### CODE REVIEW

File: lib/manager/npm/post-update/yarn.spec.ts [hunk 1]
```diff
@@ -66,9 +67,33 @@ describe(getName(), () => {
       expect(fixSnapshots(execSnapshots)).toMatchSnapshot();
     }
   );
-  it.each([['1.22.0'], ['2.1.0']])(
+  it('only skips build if skipInstalls is false', async () => {
```
Comment (viceice) [inline lib/manager/npm/post-update/yarn.spec.ts:70]: Add newline between new tests. Can be deferred for existing tests.

**Review by viceice — COMMENTED** [review]
Should also be tested on exiting yarn v1 and yarn v2 projects to be safe?

File: lib/manager/npm/post-update/yarn.spec.ts (hunk 1 above)
Comment (ylemkimon) [reply lib/manager/npm/post-update/yarn.spec.ts:70]: > Should also be tested on exiting yarn v1 and yarn v2 projects to be safe?

I tested the PR on Yarn 1 and Yarn 2 (https://github.com/ylemkimon/KaTeX).

> Can be deferred for existing tests.

What do you mean by 'deferred for existing tests'?

**Comment by rarkins** [comment]
> Where would be the best place to check this?

`skipInstalls` logic is spread a little between npm/extract and npm/post-update today, so  I think you can put it in either of those where it's useful. I guess it needs to be after we detect yarn version

File: lib/manager/npm/post-update/yarn.spec.ts (hunk 1 above)
Comment (viceice) [reply lib/manager/npm/post-update/yarn.spec.ts:70]: Only add the newlines bewteen tests you are adding / changing. But feel free to do this for all in your modified spec files. 🙃

**Comment by ylemkimon** [comment]
It now sets `skipInstalls = false` if [zero-installs](https://yarnpkg.com/features/zero-installs) is used, i.e., `.yarn/cache` and related files are detected in `extractPackageFile()`. This PR is ready for review.

File: lib/manager/npm/post-update/yarn.ts [hunk 2]
```diff
@@ -166,6 +178,10 @@ export async function generateLockFile(
       }
     });
 
+    commands = commands.map((c) =>
+      c.startsWith('yarn') ? `${c} ${cmdOptions}`.trim() : c
+    );
```
Comment (viceice) [inline lib/manager/npm/post-update/yarn.ts:183]: I don't like this change, make code more complicate

Comment (ylemkimon) [reply lib/manager/npm/post-update/yarn.ts]: Reverted.

**Review by viceice — COMMENTED** [review]
Some small changes

**Comment by viceice** [comment]
Please use re-request review buttons 🙃 
image

**Comment by ylemkimon** [comment]
I didn't know a non-member could do that :smile:

**Comment by viceice** [comment]
Also the author of the PR is allowed to use that. 🤗

File: lib/manager/npm/extract/yarn.ts [hunk 3]
```diff
@@ -36,3 +41,26 @@ export async function getYarnLock(filePath: string): Promise<LockFile> {
...
+  const paths = [conf.cacheFolder || './.yarn/cache', '.pnp.cjs', '.pnp.js'];
+  if (miscUtils.tryParseOptionalBoolean(conf.pnpEnableInlining) === false) {
+    paths.push(conf.pnpDataPath || './.pnp.data.json');
+  }
+  return paths;
+}
+
+export async function isZeroInstall(yarnrcYmlPath: string): Promise<boolean> {
+  const yarnrcYml = await readLocalFile(yarnrcYmlPath, 'utf8');
+  if (is.string(yarnrcYml)) {
```
Comment (viceice) [inline lib/manager/npm/extract/yarn.ts:56]: May use `is.nonEmptyString` ?

Comment (ylemkimon) [reply lib/manager/npm/extract/yarn.ts:56]: @viceice An empty YAML is still valid and in this case, it should check for default paths.

**Review by viceice — APPROVED** [review]

**Comment by ylemkimon** [comment]
Thank you for the reviews!

**Pull request merged**